### PR TITLE
Add register_op_hook for gluon

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##

This PR adds a register_op_hook for Block. One can register_op_hook to monitor all the ops that are called by the HybridBlock after hybridization. For example:

```
def mon_callback(node_name, opr_name, arr):
      print(py_str(node_name))
      print(py_str(opr_name))
model = mx.gluon.nn.HybridSequential(prefix="dense_")
with model.name_scope():
     model.add(mx.gluon.nn.Dense(2))
model.initialize()
model.hybridize()
model.register_op_hook(mon_callback, monitor_all=True)
model(mx.nd.ones((2, 3, 4)))
```

This will print the following : 
```
dense_dense0_fwd_data
FullyConnected
dense_dense0_fwd_weight
FullyConnected
dense_dense0_fwd_bias
FullyConnected
dense_dense0_fwd_output
FullyConnected
```

Setting monitor_all=False will print only the output : 
```
dense_dense0_fwd_output
FullyConnected
```

### What is not supported ###

1. register_op_hook on non hybridized blocks is not supported. If a block is not a hybrid block and register_op_hook is called on it, it will be registered only on HybridBlocks which are children of the blocks. The callback will only be invoked after hybridization.
2. Calback is not supported when static_shape=True for call to hybridize.

cc @Vikas-kum 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/imperative/imperative_utils.cc [hunk 1]
```diff
@@ -120,6 +121,18 @@ void InvokeOperator(const nnvm::IndexedGraph& idx,
       *arrays[eid] = NDArray();
     }
   }
+  if (callback) {
+    for (size_t i = 0; i < ndinputs.size(); ++i) {
+      NDArray* cpy = new NDArray(*ndinputs[i]);
+      callback(node.source->attrs.name.c_str(), node.source->op()->name.c_str(),
```
Comment (Vikas-kum) [inline src/imperative/imperative_utils.cc]: above you mention that cb takes one string and a array handle. 
here i see 2 strings, am i missing anything here?

File: src/imperative/imperative_utils.cc [hunk 2]
```diff
@@ -120,6 +121,18 @@ void InvokeOperator(const nnvm::IndexedGraph& idx,
...
+               reinterpret_cast<void*>(cpy));
+    }
+    for (size_t i = 0; i < ndoutputs.size(); ++i) {
+      NDArray* cpy = new NDArray(*ndoutputs[i]);
+      callback(node.source->attrs.name.c_str(), node.source->op()->name.c_str(),
```
Comment (Vikas-kum) [inline src/imperative/imperative_utils.cc]: i guess first is tensor name, 2nd is operator name, 3rd is value. Can you confirm this?
Is there a way to distinguish if the tensor is input or output ?

Also, i don't see monitor_all being used, is it used somewhere implicitly not in this PR?

File: src/imperative/imperative_utils.cc (hunk 1 above)
Comment (anirudh2290) [reply src/imperative/imperative_utils.cc]: this is still wip, the comment is wrong and needs to be cleaned up , it takes 2 strings.

File: src/imperative/imperative_utils.cc (hunk 2 above)
Comment (anirudh2290) [reply src/imperative/imperative_utils.cc]: yes you are right. there is a way to distinguish by having a postfix after node name for outputs. Alternatively we can have different callback registration apis for inputs and outputs. Currently it is common for both.

monitor_all is not being used yet and the intention was to use it for inputs vs inputs and outputs, similar to the way it is done with graph executor.

**Comment by anirudh2290** [comment]
It looks like my changed files are not being picked up and probably cached files are being picked up for CI. @marcoabreu would you know how I can force clear the cache ?

File: src/common/utils.cc [hunk 3]
```diff
@@ -51,5 +51,62 @@ void CastStorageDispatch<cpu>(const OpContext& ctx,
...
+    for (size_t i = 0; i < node->num_outputs(); ++i) {
+      output_names.emplace_back(std::to_string(i));
+    }
+  }
+
+  for (size_t i = 0; i < node->num_outputs(); ++i) {
+    if (state_arrays[idx.entry_id(nid, i)]->is_none()) {
+      continue;
+    }
+    NDArray *cpy = new NDArray(*state_arrays[idx.entry_id(nid, i)]);
```
Comment (rahul003) [inline src/common/utils.cc]: Can we avoid the copy and just pass a reference to the current array?

File: python/mxnet/gluon/block.py [hunk 4]
```diff
@@ -833,6 +848,12 @@ def _deferred_infer_shape(self, *args):
     def _call_cached_op(self, *args):
         if self._cached_op is None:
             self._build_cache(*args)
+        assert self._cached_op, "cached op is not None"
+        if self._callback:
+            self._cached_op._register_op_hook(self._callback, self._monitor_all)
+            if len(self._flags) >= 2 and self._flags[1]:
+                warnings.warn("Callback is not supported when static_shape=True "
```
Comment (rahul003) [inline python/mxnet/gluon/block.py]: Why's this? If it's not supported, do not register the callback?

Comment (rahul003) [reply python/mxnet/gluon/block.py]: else say something like experimental for static_shape=True if you are not confident, but it might work

File: src/common/utils.cc [hunk 5]
```diff
@@ -51,5 +51,62 @@ void CastStorageDispatch<cpu>(const OpContext& ctx,
...
+    }
+  }
+
+  for (size_t i = 0; i < node->num_inputs(); ++i) {
+    const nnvm::NodeEntry &input = node->inputs[i];
+    if (state_arrays[idx.entry_id(input)]->is_none()) {
+      continue;
+    }
+    NDArray *cpy = new NDArray(*state_arrays[idx.entry_id(input)]);
+    std::string name = inode.source->attrs.name + "_" + input_names[i];
```
Comment (rahul003) [inline src/common/utils.cc]: Is this naming convention standard across MXNet? For example exporting of graph in mxboard? https://github.com/awslabs/mxboard/blob/686933cfe466af324b53fcfe043093d37a0e38d1/python/mxboard/summary.py#L364

File: python/mxnet/gluon/block.py (hunk 4 above)
Comment (anirudh2290) [reply python/mxnet/gluon/block.py]: Yes I will say experimental, good point thanks. The reason this is experimental is because for certain code path the cached op directly calls engine push instead of calling InvokeOp. adding the callbacks in such cases was much more difficult since we just had the opr and the engine vars not the ndarray references. Please see line: https://github.com/apache/incubator-mxnet/blob/master/src/imperative/cached_op.cc#L689

For this case, it is not supported.

File: src/common/utils.cc (hunk 3 above)
Comment (anirudh2290) [reply src/common/utils.cc]: this wont create a copy of data, though to make changes to ndarray metadata for copy without changing the original we still need this copy.

File: src/common/utils.cc (hunk 5 above)
Comment (anirudh2290) [reply src/common/utils.cc]: I am not sure about mxboard, but in general in mxnet, the naming convention for input data doesnt prepend the node name. but since for register_op_hook we need to monitor intermediate nodes we are prepending node name here. Following the convention similar to register_monitor_callback.

**Comment by apeforest** [comment]
Have we measured how much performance overhead this might introduce?

**Comment by anirudh2290** [comment]
@apeforest no we haven't measured perf overhead since This is only used for monitoring. This should not be used in production or for perf intensive use cases.

File: src/common/utils.cc (hunk 5 above)
Comment (anirudh2290) [reply src/common/utils.cc]: I have changed it to experimental and also added check both for static_alloc and static_shape.

**Comment by anirudh2290** [comment]
Thank you @Vikas-kum, @rahul003, @apeforest for the review!!

**Comment by joapolarbear** [comment]
@anirudh2290 It seems this pr may affect the trace results, below shows the visualized results using chrome://tracing when training a bert model.
image
In contrast, before I use this pr, I get the following dense traces. Many forward and backpropagation operators miss in the above traces. 
image
FYI, I encounter the following warning when using this pr. I am not sure whether this warning has any effect.
image

**Comment by anirudh2290** [comment]
@joapolarbear do you just see much spacing out or is it that ops are missing in one but not in other. registering an op hook is expensive and called at the beginning and end of each op execution and may kill performance. It is meant to be used for debugging and should be commented out during profiling.

**Comment by eric-haibin-lin** [comment]
> registering an op hook is expensive and called at the beginning and end of each op execution and may kill performance. 


@anirudh2290 did we measure how much overhead it is to invoke a dummy hook?

**Comment by anirudh2290** [comment]
@eric-haibin-lin i didnt measure that, since this was intended only for debugging not to be used in production use cases. If I had to make a guesstimate i would say it would be very similar to invoking a dummy hook with set_monitor_callback in module.

File: python/mxnet/gluon/block.py [hunk 6]
```diff
@@ -833,6 +848,12 @@ def _deferred_infer_shape(self, *args):
     def _call_cached_op(self, *args):
         if self._cached_op is None:
             self._build_cache(*args)
+        assert self._cached_op, "cached op is not None"
+        if self._callback:
+            self._cached_op._register_op_hook(self._callback, self._monitor_all)
+            if len(self._flags) >= 2 and (self._flags[1] or self._flags[0]):
```
Comment (kovtun77) [inline python/mxnet/gluon/block.py:854]: `self._flags = list(kwargs.items())`
The condition is true even when it is not supposed to be. For example if
`self._flags = [('static_alloc', False), ('static_shape', False), ('inline_limit', 2)]`
